src/state_machine/workload.zig

```diff
@@ -948,6 +1004,28 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
             self.auditor.on_create_transfers(client_index, timestamp, transfers, results_sparse);
             if (transfers.len == 0) return;
 
+            // We don't expect batches containing both retries and new transfers.
+            // The ID cannot regress to preserve the `transfers_delivered_recently`
+            // and `transfers_delivered_past` logic.
+            if (results_sparse.len > 0 and
+                results_sparse[0].result == .id_already_failed)
+            {
+                for (results_sparse) |item| assert(item.result == .id_already_failed);
+                return;
+            }
+
+            // Enqueue the `id`s of all transient failures to be retried in the next request.
+            for (results_sparse) |item| {
+                assert(item.result != .ok);
+                assert(item.result != .id_already_failed);
+                if (item.result.transient()) {
+                    self.transient_failures.putNoClobber(
+                        transfers[item.index].id,
+                        {},
+                    ) catch unreachable;
```

Comment:
Alternatively, if the hash map had some (arbitrary, fixed, constant) capacity, then this could insert iff it is not full.